### PR TITLE
feat(IDX): use json for mainnet-canisters

### PR DESCRIPTION
**Pull request by nmattia**
This replaces `mainnet-canisters.bzl` with a json equivalent `mainnet-canisters.json`. This simplifies the update logic and separates the bazel logic from the mainnet data, also allowing mainnet data to be used more broadly if necessary. In the future this will allow us to replace the update logic with `jq` and `ic-wasm` directly.

A new repository rule is added which reads the JSON data and downloads the canisters as repositories.

### CODE REVIEW

File: mainnet-canisters.json [hunk 1]
```diff
@@ -0,0 +1 @@
...211f5770dd52e","sha256":"57c72469f01fd6ea8b5c5a962a1fed9b4ad550bbebdae38c29d5ad330c25c724"},"root":{"rev":"a0207146be211cdff83321c99e9e70baa62733c7","sha256":"c280a25dc565f8a42429cb5b969906c4c5a789381e98f6e11c247c91c4dfaac5"},"sns-wasm":{"rev":"87343a880050ca72b1361138535211f5770dd52e","sha256":"6dd00ebe425ba360be161c880ce3a3b3cda5a3738d6b323a9fd0366debf590ce"},"sns_archive":{"rev":"3d0b3f10417fc6708e8b5d844a0bac5e86f3e17d","sha256":"5c595c2adc7f6d9971298fee2fa666929711e73341192ab70804c783a0eee03f"},"sns_governance":{"rev":"87343a880050ca72b1361138535211f5770dd52e","sha256":"66156ae6686ac88836c51dc0dc970800bb83c607e83d2448c90336aa0f1ed589"},"sns_index":{"rev":"3d0b3f10417fc6708e8b5d844a0bac5e86f3e17d","sha256":"08ae5042c8e413716d04a08db886b8c6b01bb610b8197cdbe052c59538b924f0"},"sns_ledger":{"rev":"3d0b3f10417fc6708e8b5d844a0bac5e86f3e17d","sha256":"e8942f56f9439b89b13bd8037f357126e24f1e7932cf03018243347505959fd4"},"sns_root":{"rev":"87343a880050ca72b1361138535211f5770dd52e","sha256":"d52d44b6df33de56c5d02ecc8b26fbf7452df8e426d77ecc9f2e3c98a8a70316"},"swap":{"rev":"87343a880050ca72b1361138535211f5770dd52e","sha256":"4ea01d425cd9c6c0a2c4988af03710d7c2377527eb0375067c69a9baff9963f2"}}
```
Comment (basvandijk) [inline mainnet-canisters.json:1]: Any chance to have this pretty-printed such that future diffs will be shorter?

**Review by basvandijk — APPROVED** [review]

File: mainnet-canisters.json (hunk 1 above)
Comment (nmattia) [reply mainnet-canisters.json:1]: good call about the diff! we could simply replace `to_writer` to [`to_writer_pretty`](https://docs.rs/serde_json/latest/serde_json/fn.to_writer_pretty.html). Would that work for you?

File: rs/nervous_system/tools/sync-with-released-nevous-system-wasms/src/main.rs [hunk 2]
```diff
@@ -182,53 +182,32 @@ fn update_mainnet_canisters_bzl_file(
...
+        let _prev = m.insert(canister.canister_name.clone(), entry);
     }
 
     // Write the new content back to the file
-    let mut file = File::create(workspace_file_path)?;
-    for line in &new_content {
-        writeln!(file, "{}", line)?;
-    }
+    let file = File::create(canisters_json)?;
+    serde_json::to_writer(file, &m).unwrap();
```
Comment (basvandijk) [inline rs/nervous_system/tools/sync-with-released-nevous-system-wasms/src/main.rs:210]:
```suggestion
    serde_json::to_writer_pretty(file, &m).unwrap();
```

File: mainnet-canisters.json (hunk 1 above)
Comment (basvandijk) [reply mainnet-canisters.json:1]: Yes just suggested the same below

File: bazel/mainnet-canisters.bzl [hunk 3]
```diff
@@ -0,0 +1,80 @@
+"""Mainnet canister definitions.
+
+This creates a repository which exports 'canister_deps'. This macro can be
+called to create one repository for each canister in the mainnet canister list.
```
Comment (aterga) [inline bazel/mainnet-canisters.bzl:4]: "create one repository for each canister" — this is a bit confusing to me, could you provide an example?

File: WORKSPACE.bazel [hunk 4]
```diff
@@ -81,8 +81,63 @@ load("@rules_oci//oci:repositories.bzl", "oci_register_toolchains")
 oci_register_toolchains(name = "oci")
 
 load("@rules_oci//oci:pull.bzl", "oci_pull")
+load("//bazel:mainnet-canisters.bzl", "canisters")
 load("//third_party/lmdb:repository.bzl", "lmdb_repository")
-load("mainnet-canisters.bzl", "mainnet_ck_canisters", "mainnet_core_nns_canisters", "mainnet_sns_canisters")
+
+# We use inconsistent filenames and repository references, so we perform a lot
+# of remapping
```
Comment (aterga) [inline WORKSPACE.bazel:88]: Could you say a bit more in this comment: 
1. What exactly are the keys and values?
2. When should one add entries to this map, and what would happen if this maps is not consistent (e.g., due to a typo)?

**Comment by anchpop** [comment]
Note that there is not automated testing of `//rs/nervous_system/tools/sync-with-released-nevous-system-wasms`, so please run it manually to be sure it still works before merging this PR (I can also help you with any changes that need to be made)

File: WORKSPACE.bazel (hunk 4 above)
Comment (nmattia) [reply WORKSPACE.bazel]: > What exactly are the keys and values?

Let me know if the updated comment makes sense.

> When should one add entries to this map, and what would happen if this maps is not consistent (e.g., due to a typo)?

You'll get a build error early (i.e. during bazel analysis). IMO this is expected so I think not worth adding to the comment. Lmk!

File: bazel/mainnet-canisters.bzl (hunk 3 above)
Comment (aterga) [reply bazel/mainnet-canisters.bzl]: Just for my knowledge. Is whatever goes before the `//` in a target — a Bazel repository? E.g., in `@foo//hello`, `@foo` is the repository?

**Pull request merged**